components/eamxx/src/share/field/field_utils.hpp

```diff
@@ -111,6 +111,113 @@ void perturb (const Field& f,
   impl::perturb<ST>(f, engine, pdf, base_seed, level_mask, dof_gids);
 }
 
+// Utility to compute the contraction of a field along its column dimension.
+// This is equivalent to einsum('i,i...k->...k', weight, f_in); i is the column.
+// The layouts are such that:
+// - The first dimension is for the columns (COL)
+// - There can be only up to 3 dimensions
+template <typename ST>
+Field horiz_contraction(const Field &f_in, const Field *weight = nullptr,
+                        const ekat::Comm *comm = nullptr) {
+  const auto &l_in = f_in.get_header().get_identifier().get_layout();
+  const auto &n_in = f_in.get_header().get_identifier().name();
+  const auto &u_in = f_in.get_header().get_identifier().get_units();
+  const auto &g_in = f_in.get_header().get_identifier().get_grid_name();
+
+  FieldIdentifier f_out_id(n_in + "_horiz_contraction",
+                           l_in.clone().strip_dim(0), u_in, g_in);
+  // Allocate the output field
+  Field f_out(f_out_id);
+  f_out.allocate_view();
+  f_out.deep_copy(0);
+
+  // Call the implementation
+  horiz_contraction<ST>(f_out, f_in, weight, comm);
+  return f_out;
+}
+
+template <typename ST>
+void horiz_contraction(const Field &f_out, const Field &f_in,
+                       const Field *weight    = nullptr,
+                       const ekat::Comm *comm = nullptr) {
+  using namespace ShortFieldTagsNames;
+
+  const auto &l_out = f_out.get_header().get_identifier().get_layout();
+  const auto &l_in  = f_in.get_header().get_identifier().get_layout();
+  const auto &n_in  = f_in.get_header().get_identifier().name();
+  const auto &u_in  = f_in.get_header().get_identifier().get_units();
+  const auto &g_in  = f_in.get_header().get_identifier().get_grid_name();
+
+  // If weight is not provided, we set it as a field of ones
+  Field wt;
+  if(weight) {
+    wt = *weight;
+  } else {
+    FieldIdentifier wt_id(n_in + "_weight", {{COL}, {l_in.dim(0)}}, u_in, g_in);
```

Comment:
Uhm, I am not fond of this approach, since it involves a runtime allocation. Do you think it'd be too complicated to handle the case without weight without creating a dummy weight field at runtime?

Reply:
At this point we could maybe make weight mandatory. My thought was to avoid having to use a weight at all, but if it's created anyways, we may as well force the user to to it, and make sure there's no hidden runtime alloc.

Reply:
Along the line of optional-vs-mandatory args: should we make comm also a required input? I don't really foresee using this function without it...

Reply:
I'm ok forcing to pass the weight, since this is likely only going to be used for horizontal integrals, where an area weight will be needed (or, at least, that'll be the most important scenario).

As for the comm, I kind of regret making it optional in the other fcns (max/min/norm), so if that's your motivation for making it optional (consistency), I would make it mandatory.